### PR TITLE
cephadm: add --registry-{password,username} options

### PR DESCRIPTION
**Pull request by tchaikov**
allow user to log into container registry with the specified
credentials. this helps user to address the issue if the registry is
protected by authentication machinary.

Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
see also https://tracker.ceph.com/issues/45631

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -1154,6 +1155,28 @@ def default_image(func):
...
+            out, err, code = call([container_path, 'login',
+                                   '-u', args.registry_username,
+                                   '-p', args.registry_password])
+            if code != 0:
+                logger.warning('unable to login: %s', err)
+                self.do_login = False
+
+    def __exit__(self):
+        if self.do_login:
+            call([container_path, 'logout'])
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:1178]: just wondering. would something like this also work?

```python
def container_registry_credentials(func):
    do_login = ...
    def wrapper(*args, *+kwargs):
        try:
            ... = call([container_path, 'login', ...
            func(*args, **kwargs)
        finally:
            call([container_path, 'logout'])
    return wrapper
```
?

**Comment by sebastian-philipp** [comment]
mypy: 
```
==================================== ERRORS ====================================
____________________ ERROR collecting tests/test_cephadm.py ____________________
tests/test_cephadm.py:11: in <module>
    cd = SourceFileLoader('cephadm', 'cephadm').load_module()
cephadm:2220: in <module>
    @container_registry_credentials()
cephadm:1160: in __init__
    if not args.registry_username:
E   NameError: name 'args' is not defined
```

thanks for the PR! Just a few minor questions:

1. do we need mgr/cephadm module options as well? otherwise no one will add the arguments?
2. we we need that in the unit files as well? I mean, `podman run` might also pull befor running the container.
3. fixes https://tracker.ceph.com/issues/44886 as well?
4. do you know https://www.mankier.com/5/containers-auth.json ?

File: src/cephadm/cephadm (hunk 1 above)
Comment (tchaikov) [reply src/cephadm/cephadm:1178]: This is actually my first version. Was using context manager. I think it’d be nice to use it as a decorator. 

If you think it’s an overdesign. I will use a this approach.

**Comment by tchaikov** [comment]
I was not sure if it was an acceptable fix when working on this last night. But seems you are interested in this. And it’s actually a feature!

Yes, we need a way to pass these options down when cephadm is called by mgr/cephadm.

Agreed. But I like the cached registry better.

**Comment by tchaikov** [comment]
> mypy:
> 
> ```
> ==================================== ERRORS ====================================
> ____________________ ERROR collecting tests/test_cephadm.py ____________________
> tests/test_cephadm.py:11: in <module>
>     cd = SourceFileLoader('cephadm', 'cephadm').load_module()
> cephadm:2220: in <module>
>     @container_registry_credentials()
> cephadm:1160: in __init__
>     if not args.registry_username:
> E   NameError: name 'args' is not defined
> ```
> 
> thanks for the PR! Just a few minor questions:
> 
> 1. do we need mgr/cephadm module options as well? otherwise no one will add the arguments?

Yes, we need a way to pass these options down when cephadm is called by mgr/cephadm.

> 2. we we need that in the unit files as well? I mean, `podman run` might also pull befor running the container.

Agreed. But I like the cached registry better.

> 3. fixes https://tracker.ceph.com/issues/44886 as well?

I was not sure if it was an acceptable fix when working on this last night. But seems you are interested in this. And it’s actually a feature!

> 4. do you know https://www.mankier.com/5/containers-auth.json ?

i know. that's why i want to logout to avoid security issues by calling "logout". do you have better idea how to keep this caches credentials around without hurting the performance?

**Comment by tchaikov** [comment]
> 
> > 1. do you know https://www.mankier.com/5/containers-auth.json ?
> 
> i know. that's why i want to logout to avoid security issues by calling "logout". do you have better idea how to keep this cached credentials around without hurting the performance? how about having a global instance of this decorator / context manager, and call "logout" in its destructor?

after a second thought, i am not sure about anymore. as "logout" command nukes **all** the cached credentials. no matter if they are added by cephadm or not.

but i am keeping it anyway. if you think i should remove it, i will do.

**Comment by sebastian-philipp** [comment]
Caching within cephadm itself won't help.

1. There is typically only **one** container created per invocation. (the exception is bootstrap, which creates two containers in a single invocation) 
2. deploying a daemon doesn't pull anything: pulling the containers is executed by `podman ru in the context of the systemd unit.
3. There is no such thing as a 'global instance'. `src/cephadm` is just a CLI tool without any persistent state.

Also, how do you make `cephadm shell` work without adding the registry credentials all the time?

Which makes me think. Having a solution where `podman run` just works ootb would be my preferred way to get this done.

**Comment by sebastian-philipp** [comment]
https://pad.ceph.com/p/cephadm-registry-credentials

**Comment by sebastian-philipp** [comment]
still this PR makes sense, if it is done right. @tchaikov you want to continue working on this?

**Comment by tchaikov** [comment]
this PR was created to address the test failures. since it has been fixed by you, its priority is now lowered. i will come back once i have more bandwidth.

**Pull request closed without merging**